### PR TITLE
First draft of did-pkh method.

### LINKED ISSUE #1024: [proposal] add did:pkh support
### Requirements
* [x] inline resolver; we can use the resolver from ceramic: [pkh-did-resolver](https://github.com/ceramicnetwork/js-ceramic/tree/develop/packages/pkh-did-resolver)
>```ts
> import { getResolver as getPkhResolver } from 'pkh-did-resolver'
> // ...
> new DIDResolverPlugin({
>   // ...
>   ...getPkhResolver(),
> })
> ```
* [ ] `@veramo/did-provider-pkh` - at least for ethereum
This requires an implementation of [`AbstractIdentifierProvider`](https://github.com/uport-project/veramo/blob/next/packages/did-manager/src/abstract-identifier-provider.ts)
Since did:pkh doesn't support updates, this provider would be very similar to [`@veramo/did-provider-key`](https://github.com/uport-project/veramo/blob/0c05a662e93def87f3b6750bbb35b79996e3ee6c/packages/did-provider-key/src/key-did-provider.ts#L16), in the sense that most methods would be stubs, and that DIDs are created from a key.

The first implementation can support only the ethereum based DIDs, `did:pkh:eip155:...`.
The parameters for creating a DID linked to a particular network should be [inspired by `@veramo/did-provider-ethr`](https://github.com/uport-project/veramo/blob/0c05a662e93def87f3b6750bbb35b79996e3ee6c/packages/did-provider-ethr/src/ethr-did-provider.ts#L34) (the same file has a lot of other code that should prove useful).

### testing

The PR closing this issue should contain a test where a `did:pkh` is created and used as an issuer in a call to `createVerifiableCredential()`, then that credential should be verified successfully using `verifyCredential()`.
It is acceptable to provide only integration tests for this (meaning tests added to the top level [`__tests__` folder](https://github.com/uport-project/veramo/tree/next/__tests__))

### constraints

* **Environments**: All plugin packages published in this repository are meant to work in nodejs, web(react) and mobile(react-native) environments. This will apply to this new package too.
* **No new dependencies**: Packages in this repository are often installed together so dependencies must be reused between them when possible. No new dependencies should be introduced if this can be avoided.

### PR DESCRIPTION
**Pull request by dchagastelles**
- Create new pkh-did-provider
The pkh-did-provider supports only eip155 network id (did:pkh:eip155:...) so for now the network id is hardcoded, but we could easily parameterize to support other networks.

- Add resolve did-pkh tests
Tests passing 

- Add create did-pkh identifier test
Tests passing

- Add VC creation/verification using did-pkh identifier
Tests passing 

## Detail
Currently not using pkh-did-resolver package (error import). Code momentarily copied to did-provider-pkh resolver.

## What issue is this PR fixing
Resolves #1024 

## Quality
Check all that apply:
* [X] I want these changes to be integrated
* [x] I successfully ran `yarn`, `yarn build`, `yarn test`, `yarn test:browser` locally.
* [X] I allow my PR to be updated by the reviewers (to speed up the review process).
* [ ] I added unit tests. 
* [X] I added integration tests.

### CODE REVIEW

**Review by mirceanis — CHANGES_REQUESTED** [review]
Thank you for contributing this and apologies for taking so long to review.

I made some suggestions to fix the failing test.
The DID string for eip155 is formed using the ethereum address, not the compressed public key. This was also the cause of the test failures you were seeing.

Also, CAIP10/eip155 require chainIDs to be used in the `blockchainAccountId`, not network names.
`eip155:1:0xaddress` instead of `eip155:mainnet:0xaddress`

File: __tests__/shared/verifiableDataLD.ts [hunk 1]
```diff
@@ -257,6 +260,49 @@ export default (testContext: {
...
+      expect(verifiableCredential).toHaveProperty('proof.jws')
+      // expect(verifiableCredential.proof.verificationMethod).toEqual(
+      //   `${didKeyIdentifier.did}#${didKeyIdentifier.did.substring(
+      //     didKeyIdentifier.did.lastIndexOf(':') + 1,
+      //   )}`,
+      // )
+
+      expect(verifiableCredential['@context']).toEqual([
+        'https://www.w3.org/2018/credentials/v1',
+        'https://veramo.io/contexts/profile/v1',
```
Comment (mirceanis) [inline __tests__/shared/verifiableDataLD.ts:293]:
```suggestion
        'https://veramo.io/contexts/profile/v1',
        'https://w3id.org/security/suites/secp256k1recovery-2020/v2',
```

File: __tests__/shared/verifiableDataLD.ts [hunk 2]
```diff
@@ -257,6 +260,49 @@ export default (testContext: {
...
+      ])
+      expect(verifiableCredential['type']).toEqual(['VerifiableCredential', 'Profile'])
+
+      storedPkhCredentialHash = await agent.dataStoreSaveVerifiableCredential({ verifiableCredential })
+      expect(typeof storedPkhCredentialHash).toEqual('string')
+
+      const verifiableCredential2 = await agent.dataStoreGetVerifiableCredential({
+        hash: storedCredentialHash,
```
Comment (mirceanis) [inline __tests__/shared/verifiableDataLD.ts:301]:
```suggestion
        hash: storedPkhCredentialHash,
```

**Review by mirceanis — CHANGES_REQUESTED** [review]
I added another couple of suggested fixes to the tests.
Please take a look.

File: packages/did-provider-pkh/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Change Log
```
[the remark this answers is not in the document]
Comment (dchagastelles) [reply packages/did-provider-pkh/CHANGELOG.md]: done.

File: packages/did-provider-pkh/src/pkh-did-provider.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    // if (typeof networkSpecifier === 'number') {
+    //   networkSpecifier =
+    //     network.name && network.name.length > 0
+    //       ? network.name
+    //       : BigNumber.from(options?.network || 1).toHexString()
+    // }
+    //const networkString = networkSpecifier && networkSpecifier !== 'mainnet' ? `${networkSpecifier}:` : ''
+    const identifier: Omit<IIdentifier, 'provider'> = {
+      did: 'did:pkh:eip155:' + network + ':' + compressedPublicKey,
```
[the remark this answers is not in the document]
Comment (dchagastelles) [reply packages/did-provider-pkh/src/pkh-did-provider.ts]: Used toEthereumAddress helper to obtain EthereumAddress directly from publickey hex.

File: packages/did-provider-ion/dist/tsdoc-metadata.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// This file is read by tools that parse documentation comments conforming to the TSDoc standard.
```
[the remark this answers is not in the document]
Comment (dchagastelles) [reply packages/did-provider-ion/dist/tsdoc-metadata.json]: removed

**Comment by nickreynolds** [comment]
now that tests are passing, it would probably be a good time to rebase this branch against `next` to get rid of the extraneous commits

File: packages/did-provider-pkh/package.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    "@ethersproject/signing-key": "^5.7.0",
+    "@ethersproject/bignumber": "^5.7.0",
+    "@ethersproject/transactions": "^5.7.0",
+    "@veramo/core": "^4.0.0",
+    "@veramo/did-manager": "^4.0.0",
+    "debug": "^4.3.3"
+  },
+  "devDependencies": {
+    "@types/debug": "4.1.7",
+    "typescript": "4.7.3"
```
Comment (mirceanis) [inline packages/did-provider-pkh/package.json:21]: align with existing versions
```suggestion
    "typescript": "4.8.4"
```

**Review by mirceanis — APPROVED** [review]
Looks good.
Thanks for contributing this!

File: package.json [hunk 7]
```diff
@@ -56,6 +56,7 @@
     "rimraf": "3.0.2",
     "semantic-release": "19.0.3",
     "ts-jest": "28.0.4",
+    "caip": "^1.1.0"
```
Comment (mirceanis) [inline package.json:59]:
```suggestion
    "caip": "^1.1.0",
```

**Review by mirceanis — APPROVED** [review]

**Pull request merged**